### ISSUE TITLE
Document Type References In Docs Site

### ISSUE THREAD

**Issue by shellscape** [rollup/rollup#3326]
Documentation Is:

- [x] Missing
- [x] Needed
- [ ] Confusing
- [ ] Not Sure?

### Please Explain in Detail...

We should add a section to the documentation website to cover types that are mentioned in the documentation. `RollupWarning` is the type that caught my eye on this, as it has many properties on its own and inherited from `RollupLogProps`. I had looked to the source for an explanation of properties and how they were used, but the source contains no annotations: https://github.com/rollup/rollup/blob/b99758b3a4617769f2371540e530536d5c246692/src/rollup/types.d.ts#L12-L25

### Your Proposal for Changes

I propose we go through the docs as they are and list the types mentioned there. From that list, we should add a section that describes the types in detail. For example:

- How is `RollupWarning.plugin` used and how does it affect console output?
- How does `RollupWarning.code` vary from `RollupWarning.pluginCode`?
- What are the best practices for how to use the properties of `RollupWarning` and `RollupError` in plugin development?

We're seeing an uptick in interest in both Rollup and in plugin development (both core and third party) and I'd love to encourage best practices, as well as know them myself.

**Comment by lukastaegert** [comment]
You are very right this should be documented better. Here is a quick summary, hope we can polish it and put it into the docs soon:

- `plugin` is actually added by Rollup to designate which plugin caused a warning or error i.e. it does not need to be handled by plugins.
- for warnings from plugins, `code` is always `PLUGIN_WARNING`. If a plugin adds a `code` to identify a specific warning, Rollup copies that code to `pluginCode`

This is all handled here: https://github.com/rollup/rollup/blob/b99758b3a4617769f2371540e530536d5c246692/src/utils/PluginContext.ts#L178-L184

There are indeed a few properties that plugins can use that Rollup CLI will pick up to make warnings and errors nicer. For warnings, this is handled here:
https://github.com/rollup/rollup/blob/b99758b3a4617769f2371540e530536d5c246692/cli/run/batchWarnings.ts#L174-L198

So these are the properties that can be used:
- `url`: A URL with further information. Very nice if warnings have one of those so users can learn more details what they should change.
- `id`: If the warning is caused by something that happens in a specific module, this should be the resolved `id` of the module. The CLI will again display a shortened version of the `id` relative to the current working directory.
- `loc`: Only if `id` is present, this is a `{line: number, column: number}` object that will be displayed with the file name if present to designate a specific code location.
- `frame`: A piece of text to be displayed below the warning, usually to show the code surrounding an error.

Additionally only in `transform` hooks, there is some advanced logic that plugins can use:
https://github.com/rollup/rollup/blob/b99758b3a4617769f2371540e530536d5c246692/src/utils/transform.ts#L118-L124

Basically you can specify a `pos` property which is just a number corresponding to a position in a file, in which case Rollup will automatically generate `id`, `loc` and `frame` for you.

For errors, the logic is very similar.

**Comment by lukastaegert** [comment]
As for the other properties of `RollupLogProps`, most are specific to a single or few errors or warnings and it would only make sense to document them as part of a list describing possible errors and warnings. Which would be a huge piece of work, not sure it is worth it.
